File: python/src/schemas.py
"""
Week 2 — Structured Outputs

Pydantic models + structured LLM calls.
The LLM returns typed objects, not prose.

Two schema families:
  BuildCheck            → flat, 4-field PR analysis (in-session)
  ServiceReadinessReport → composed, nested, validated (self-learning + capstone preview)
"""
from datetime import datetime
from langchain_core.messages import HumanMessage, SystemMessage
from pydantic import BaseModel, Field, field_validator, model_validator
from typing import Literal, Optional

from src.llm import get_llm


# ═══════════════════════════════════════════════════════════════
# PR Analysis — flat schema for the in-session exercise
# ═══════════════════════════════════════════════════════════════

class BuildCheck(BaseModel):
    """A structured analysis of a PR or code change."""
    project: str = Field(description="Project or service name")
    severity: Literal["low", "medium", "high", "critical"] = Field(
        description="How urgent is this change?"
    )
    summary: str = Field(description="One-sentence summary of the change")
    affected_files: list[str] = Field(description="Files modified by this change")


# ═══════════════════════════════════════════════════════════════
# Service Readiness Report — composed schema for self-learning
#
# This mirrors the DevBuddy end-state output.
# Engineers build and test this with mock data — no API calls.
#
# Architecture (from Plan.md):
#   User Query → Guardrail → Orchestrator → RAG + Tools → Structured Output
#   Final output: {ready, blockers, next_steps, evidence}
# ═══════════════════════════════════════════════════════════════

class ServiceInfo(BaseModel):
    """Identity of the service being evaluated."""
    name: str = Field(description="Service name, e.g. 'auth-service'")
    version: str = Field(description="Currently deployed version, e.g. '2.1.0'")
    owner_team: str = Field(description="Team responsible for this service")


class BuildStatus(BaseModel):
    """Current build health of the service."""
    status: Literal["healthy", "degraded", "down", "unknown"] = Field(
        description="Current build/health status"
    )
    last_deploy: str = Field(description="ISO timestamp of the most recent deployment")
    failing_since: Optional[str] = Field(
        None,
        description="ISO timestamp of when the build started failing, if degraded or down"
    )

    @field_validator("failing_since")
    @classmethod
    def failing_since_required_for_unhealthy(cls, v, info):
        """If status is degraded or down, failing_since should be present."""
        status = info.data.get("status")
        if status in ("degraded", "down") and v is None:
            raise ValueError(
                f"failing_since is required when status is '{status}'. "
                f"Provide an ISO timestamp."
            )
        return v


class DeployRecord(BaseModel):
    """A single deployment event."""
    sha: str = Field(description="Commit SHA of the deployment")
    author: str = Field(description="Engineer who triggered the deploy")
    timestamp: str = Field(description="ISO timestamp of the deployment")
    status: Literal["success", "failed", "rolling_back"] = Field(
        description="Outcome of this deployment"
    )


class DeploymentInfo(BaseModel):
    """Recent deployment history for the service."""
    recent_deploys: list[DeployRecord] = Field(
        description="Last N deployments, most recent first"
    )
    active_incidents: list[str] = Field(
        default_factory=list,
        description="IDs or summaries of any active incidents"
    )


class EvidenceChunk(BaseModel):
    """A piece of supporting context retrieved or produced during analysis."""
    source: Literal["rag", "tool", "user"] = Field(
        description="Where this evidence came from — RAG retrieval, tool output, or user query"
    )
    content: str = Field(description="The evidence content")
    relevance_score: Optional[float] = Field(
        None,
        description="0.0–1.0 relevance score from the retriever, if sourced from RAG"
    )


class ReadinessVerdict(BaseModel):
    """Final verdict: is this service ready for the target version?"""
    ready: bool = Field(description="True if the service can proceed to the target version")
    confidence: Literal["low", "medium", "high"] = Field(
        description="How confident is this verdict?"
    )
    blockers: list[str] = Field(
        default_factory=list,
        description="Reasons the service is NOT ready. Empty if ready=true."
    )
    recommended_next_steps: list[str] = Field(
        default_factory=list,
        description="What to do next — regardless of ready/blocked"
    )


class ServiceReadinessReport(BaseModel):
    """
    The structured output DevBuddy produces for a service readiness check.

    This is what the capstone delivers. It composes context (RAG),
    live data (tools), and reasoning into one typed, validated contract.

    Engineers test this with mock data during Week 2 self-learning.
    """
    service: ServiceInfo
    build: BuildStatus
    deployment: DeploymentInfo
    verdict: ReadinessVerdict
    evidence: list[EvidenceChunk] = Field(
        default_factory=list,
        description="Supporting evidence from retrieval and tool calls"
    )

    @model_validator(mode="after")
    def readiness_sanity_check(self):
        """
        Cross-field invariant: if ready=True, blockers must be empty.
        If ready=False, blockers should be non-empty (confidence='low' allows
        missing blockers — that's the 'we don't know' case).
        """
        if self.verdict.ready and self.verdict.blockers:
            raise ValueError(
                f"Contradiction: verdict.ready=True but blockers={self.verdict.blockers}. "
                f"If there are blockers, ready must be False."
            )
        if not self.verdict.ready and self.verdict.confidence != "low" and not self.verdict.blockers:
            raise ValueError(
                f"verdict.ready=False with confidence='{self.verdict.confidence}' "
                f"but no blockers listed. Either lower confidence to 'low' or add blockers."
            )
        return self

    @model_validator(mode="after")
    def confidence_low_if_no_evidence(self):
        """If there's no evidence, confidence cannot be higher than 'low'."""
        if not self.evidence and self.verdict.confidence != "low":
            raise ValueError(
                f"No evidence provided but confidence='{self.verdict.confidence}'. "
                f"Without evidence, confidence must be 'low'."
            )
        return self


def analyze_pr(
    title: str,
    diff: str,
    temperature: float = 0.0,
    max_tokens: int | None = None,
) -> BuildCheck:
    """
    Analyze a PR and return a structured BuildCheck.

    Args:
        title: PR title
        diff: PR diff content
        temperature: 0.0 for deterministic output (default)
        max_tokens: Max tokens in the response. None = model default.

    Returns:
        BuildCheck with project, severity, summary, affected_files
    """
    llm = get_llm(temperature=temperature, max_tokens=max_tokens)
    structured = llm.with_structured_output(BuildCheck)

    result = structured.invoke([
        SystemMessage(content=(
            "You are a code reviewer. Analyze the given PR and return a BuildCheck.\n"
            "- severity: 'critical' if it touches auth, payments, or security. "
            "'high' if it changes core logic. 'medium' for feature work. 'low' for docs/typos.\n"
            "- summary: one sentence describing what changed and why.\n"
            "- affected_files: list the files mentioned in the diff.\n"
            "- project: extract the project or service name from the PR context."
        )),
        HumanMessage(content=(
            f"PR Title: {title}\n\n"
            f"Diff:\n{diff}"
        )),
    ])

    return result

File: nodejs/src/schemas.js
/**
 * Week 2 — Structured Outputs (schemas only)
 *
 * Zod schemas — the *contract*. No LLM imports here, so these are pure and
 * testable without a network or an API key.
 *
 * Two schema families:
 *   BuildCheck            → flat, 4-field PR analysis (in-session)
 *   ServiceReadinessReport → composed, nested, validated (self-learning + capstone preview)
 *
 * The functions that call the LLM (analyzePr, generateReadinessReport) live in
 * llm_functions.js — the "action" that fills this contract.
 */
import { z } from "zod";

// ═══════════════════════════════════════════════════════════════
// PR Analysis — flat schema for the in-session exercise
// ═══════════════════════════════════════════════════════════════

export const BuildCheckSchema = z.object({
  project: z.string().describe("Project or service name"),
  severity: z
    .enum(["low", "medium", "high", "critical"])
    .describe("How urgent is this change?"),
  summary: z.string().describe("One-sentence summary of the change"),
  affected_files: z
    .array(z.string())
    .describe("Files modified by this change"),
});

/**
 * @typedef {z.infer<typeof BuildCheckSchema>} BuildCheck
 */

// ═══════════════════════════════════════════════════════════════
// Service Readiness Report — composed schema for self-learning
//
// This mirrors the DevBuddy end-state output.
// Engineers build and test this with mock data — no API calls.
// ═══════════════════════════════════════════════════════════════

export const ServiceInfoSchema = z.object({
  name: z.string().describe("Service name, e.g. 'auth-service'"),
  version: z
    .string()
    .describe("Currently deployed version, e.g. '2.1.0'"),
  owner_team: z.string().describe("Team responsible for this service"),
});

export const BuildStatusSchema = z
  .object({
    status: z
      .enum(["healthy", "degraded", "down", "unknown"])
      .describe("Current build/health status"),
    last_deploy: z
      .string()
      .describe("ISO timestamp of the most recent deployment"),
    failing_since: z
      .string()
      .nullable()
      .optional()
      .describe(
        "ISO timestamp of when the build started failing, if degraded or down"
      ),
  })
  .refine(
    (data) => {
      // If status is degraded or down, failing_since should be present
      if (
        (data.status === "degraded" || data.status === "down") &&
        !data.failing_since
      ) {
        return false;
      }
      return true;
    },
    {
      message:
        "failing_since is required when status is 'degraded' or 'down'. Provide an ISO timestamp.",
      path: ["failing_since"],
    }
  );

export const DeployRecordSchema = z.object({
  sha: z.string().describe("Commit SHA of the deployment"),
  author: z.string().describe("Engineer who triggered the deploy"),
  timestamp: z.string().describe("ISO timestamp of the deployment"),
  status: z
    .enum(["success", "failed", "rolling_back"])
    .describe("Outcome of this deployment"),
});

export const DeploymentInfoSchema = z.object({
  recent_deploys: z
    .array(DeployRecordSchema)
    .describe("Last N deployments, most recent first"),
  active_incidents: z
    .array(z.string())
    .default([])
    .describe("IDs or summaries of any active incidents"),
});

export const EvidenceChunkSchema = z.object({
  source: z
    .enum(["rag", "tool", "user"])
    .describe(
      "Where this evidence came from — RAG retrieval, tool output, or user query"
    ),
  content: z.string().describe("The evidence content"),
  relevance_score: z
    .number()
    .min(0)
    .max(1)
    .nullable()
    .optional()
    .describe(
      "0.0–1.0 relevance score from the retriever, if sourced from RAG"
    ),
});

export const ReadinessVerdictSchema = z.object({
  ready: z
    .boolean()
    .describe("True if the service can proceed to the target version"),
  confidence: z
    .enum(["low", "medium", "high"])
    .describe("How confident is this verdict?"),
  blockers: z
    .array(z.string())
    .default([])
    .describe("Reasons the service is NOT ready. Empty if ready=true."),
  recommended_next_steps: z
    .array(z.string())
    .default([])
    .describe("What to do next — regardless of ready/blocked"),
});

export const ServiceReadinessReportSchema = z
  .object({
    service: ServiceInfoSchema,
    build: BuildStatusSchema,
    deployment: DeploymentInfoSchema,
    verdict: ReadinessVerdictSchema,
    evidence: z
      .array(EvidenceChunkSchema)
      .default([])
      .describe("Supporting evidence from retrieval and tool calls"),
  })
  .refine(
    (data) => {
      // If ready=true, blockers must be empty
      if (data.verdict.ready && data.verdict.blockers.length > 0) {
        return false;
      }
      return true;
    },
    {
      message:
        "Contradiction: verdict.ready=true but blockers are present. " +
        "If there are blockers, ready must be false.",
      path: ["verdict", "ready"],
    }
  )
  .refine(
    (data) => {
      // If ready=false with confidence != 'low', blockers should be non-empty
      if (
        !data.verdict.ready &&
        data.verdict.confidence !== "low" &&
        data.verdict.blockers.length === 0
      ) {
        return false;
      }
      return true;
    },
    {
      message:
        "verdict.ready=false with confidence above 'low' but no blockers listed. " +
        "Either lower confidence to 'low' or add blockers.",
      path: ["verdict", "confidence"],
    }
  )
  .refine(
    (data) => {
      // If no evidence, confidence cannot be higher than 'low'
      if (data.evidence.length === 0 && data.verdict.confidence !== "low") {
        return false;
      }
      return true;
    },
    {
      message:
        "No evidence provided but confidence is above 'low'. " +
        "Without evidence, confidence must be 'low'.",
      path: ["verdict", "confidence"],
    }
  );

/**
 * @typedef {z.infer<typeof ServiceReadinessReportSchema>} ServiceReadinessReport
 */
